File: typescript/behavioral/observer/20210201/publisher.ts
import {Subscriber} from "./subscriber";

export class Publisher {
  private subscribers: Subscriber[] = []
  subscribe(s: Subscriber): void {
    this.subscribers.push(s)
  }
  unsubscribe(s: Subscriber): void {
    const index = this.subscribers.indexOf(s)
    this.subscribers.splice(index, 1)
  }
  notify() {
    this.subscribers.forEach(v => {
      v.update(this)
    })
  }
}

File: dart/behavioral/observer/20210201ps/bin/publisher.dart
import 'subscriber.dart';

class Publisher {
  List<Subscriber> _subscribers = [];

  void subscribe(Subscriber s) {
    _subscribers.add(s);
  }

  void unsubscribe(Subscriber s) {
    _subscribers.remove(s);
  }

  void notify() {
    _subscribers.forEach((v) {
      v.update(this);
    });
  }
}
